### PR TITLE
coral-hive and coral-trino: Parse and unparse "INTERVAL '3' DAY"

### PR DESCRIPTION
**Pull request by uzshao**
This allows coral to support queries like `CAST('2021-08-31' AS DATE) + INTERVAL '3' DAY`

### CODE REVIEW

**Comment by funcheetah** [comment]
Thanks @uzshao for this patch. Could you please add description and rebase latest master so that we can run regression tests? Thanks.

**Comment by uzshao** [comment]
@funcheetah Added description and rebased on master

**Comment by funcheetah** [comment]
Thanks. Started integration tests.

File: coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/AbstractASTVisitor.java [hunk 1]
```diff
@@ -282,6 +282,17 @@ protected R visit(ASTNode node, C ctx) {
       case HiveParser.TOK_WINDOWVALUES:
         return visitWindowValues(node, ctx);
 
+      // See IdentifierParser.g:
```
Comment (ljfgem) [inline coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/AbstractASTVisitor.java:285]: nit: `IdentifiersParser.g`

File: coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/AbstractASTVisitor.java [hunk 2]
```diff
@@ -282,6 +282,17 @@ protected R visit(ASTNode node, C ctx) {
...
+      case HiveParser.TOK_INTERVAL_DAY_LITERAL:
+      case HiveParser.TOK_INTERVAL_DAY_TIME_LITERAL:
+      case HiveParser.TOK_INTERVAL_HOUR_LITERAL:
+      case HiveParser.TOK_INTERVAL_MINUTE_LITERAL:
+      case HiveParser.TOK_INTERVAL_MONTH_LITERAL:
+      case HiveParser.TOK_INTERVAL_SECOND_LITERAL:
+      case HiveParser.TOK_INTERVAL_YEAR_LITERAL:
+      case HiveParser.TOK_INTERVAL_YEAR_MONTH_LITERAL:
```
Comment (ljfgem) [inline coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/AbstractASTVisitor.java:293]: There are also `TOK_INTERVAL_DAY_TIME` and `TOK_INTERVAL_YEAR_MONTH`, do we need to add them?

**Review by ljfgem — COMMENTED** [review]
Thanks @uzshao for this patch!
Just several nits, and could you please add the tests for Coral-Spark?

File: coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/AbstractASTVisitor.java (hunk 2 above)
Comment (uzshao) [reply coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/AbstractASTVisitor.java:293]: I did try to look for those but couldn't find how those are supported / generated.  Also Apache Calcite doesn't support DAY_TIME and YEAR_MONTH parsing yet.

Also I haven't seen a real use case for these two in practice.

So I guess it's OK to leave them out until a real use case comes up?

Comment (ljfgem) [reply coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/AbstractASTVisitor.java:293]: Makes sense, thank you for explaining.

**Review by uzshao — COMMENTED** [review]
Addressed review comments.  Added a test case for `CoralSpark`.

File: coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java [hunk 3]
```diff
@@ -175,6 +175,12 @@ public void testViews(String database, String view, String expectedSql) {
 
         { "test", "pmod_view", "SELECT MOD(MOD(- 9, 4) + 4, 4)\nFROM \"test\".\"tablea\"" },
 
-        { "test", "nullscollationd_view", "SELECT \"a\", \"b\", \"c\"\nFROM \"test\".\"tabler\"\nORDER BY \"b\" DESC" }, };
+        { "test", "nullscollationd_view", "SELECT \"a\", \"b\", \"c\"\nFROM \"test\".\"tabler\"\nORDER BY \"b\" DESC" },
+
+        { "test", "view_with_date_and_interval", "SELECT (CAST('2021-08-30' AS DATE) + INTERVAL '3' DAY)\nFROM \"test\".\"tablea\"" },
```
Comment (wmoustafa) [inline coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java:180]: Can we add a test case to more complex ones like `TOK_INTERVAL_YEAR_MONTH_LITERAL`?

File: coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java [hunk 4]
```diff
@@ -175,6 +175,12 @@ public void testViews(String database, String view, String expectedSql) {
...
+
+        { "test", "view_with_timestamp_and_interval", "SELECT (CAST('2021-08-30' AS TIMESTAMP) + INTERVAL '3' MONTH)\nFROM \"test\".\"tablea\"" },
```
Comment (wmoustafa) [inline coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java:182]: Can we test negative values?

File: coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/ParseTreeBuilder.java [hunk 5]
```diff
@@ -1029,6 +1031,43 @@ protected SqlNode visitTableTokOrCol(ASTNode node, ParseContext ctx) {
     return visitChildren(node, ctx).get(0);
   }
 
+  private SqlIntervalQualifier fromASTIntervalTypeToSqlIntervalQUalifier(ASTNode node) {
```
[the remark this answers is not in the document]
Comment (uzshao) [reply coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/parsetree/ParseTreeBuilder.java]: Good catch.  Fixed!

File: coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java (hunk 3 above)
Comment (uzshao) [reply coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java:181]: Added

**Comment by uzshao** [comment]
@wmoustafa and @ljfgem   This patch is complete now with added tests that covers negative cases, DAY TO SECOND, YEAR TO MONTH cases.  I also added support for unparsing in coral-spark.

**Review by funcheetah — APPROVED** [review]

**Pull request merged**